clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -243,12 +243,13 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, C
                     while (batchIterator.hasNext()) {
                         RecordBatch batch = batchIterator.next();
                         boolean isFull = batch != lastBatch || batch.isFull();
-                        // check if the batch is expired
-                        if (batch.maybeExpire(requestTimeout, retryBackoffMs, now, this.lingerMs, isFull)) {
+                        // Check if the batch has expired. Actual expiry is done after completing the iterations
```

Comment:
Maybe we should clarify that we invoke the callback and deallocate after the iterations (because we close the batch here, one could say that we have done part of what one may consider expiration).